api_core/google/api_core/protobuf_helpers.py

```diff
@@ -76,3 +82,150 @@ def get_messages(module):
         if inspect.isclass(candidate) and issubclass(candidate, Message):
             answer[name] = candidate
     return answer
+
+
+def _resolve_subkeys(key, separator='.'):
+    """Given a key which may actually be a nested key, return the top level
+    key and any nested subkeys as separate values.
+
+    Args:
+        key (str): A string that may or may not contain the separator.
+        separator (str): The namespace separator. Defaults to `.`.
+
+    Returns:
+        Tuple[str, str]: The key and subkey(s).
+    """
+    subkey = None
+    if separator in key:
+        index = key.index(separator)
+        subkey = key[index + 1:]
+        key = key[:index]
+    return key, subkey
+
+
+def get(msg_or_dict, key, default=_SENTINEL):
+    """Retrieve a key's value from a protobuf Message or dictionary.
+
+    Args:
+        mdg_or_dict (Union[~google.protobuf.message.Message, Mapping]): the
+            object.
+        key (str): The key to retrieve from the object.
+        default (Any): If the key is not present on the object, and a default
+            is set, returns that default instead. A type-appropriate falsy
+            default is generally recommended, as protobuf messages almost
+            always have default values for unset values and it is not always
+            possible to tell the difference between a falsy value and an
+            unset one. If no default is set then raises :class:`KeyError` will
+            be raised if the key is not present in the object.
+
+    Returns:
+        Any: The return value from the underlying Message or dict.
+
+    Raises:
+        KeyError: If the key is not found. Note that, for unset values,
+            messages and dictionaries may not have consistent behavior.
+        TypeError: If ``msg_or_dict`` is not a Message or Mapping.
+    """
+    # We may need to get a nested key. Resolve this.
+    key, subkey = _resolve_subkeys(key)
+
+    # Attempt to get the value from the two types of objects we know about.
+    # If we get something else, complain.
+    if isinstance(msg_or_dict, Message):
+        answer = getattr(msg_or_dict, key, default)
```

Comment:
There is no better way to do this with the `google.protobuf` API? Something like `HasField` (not the right choice here, but a relative maybe)?

Reply:
I was not able to find one when I wrote this (months ago).

Reply:
What about something like this:

```python
>>> from google.cloud.bigtable._generated import data_pb2
>>>
>>> t1 = data_pb2.TimestampRange()
>>> t2 = data_pb2.TimestampRange(start_timestamp_micros=100)
>>> t3 = data_pb2.TimestampRange(end_timestamp_micros=200)
>>> t4 = data_pb2.TimestampRange(
...     start_timestamp_micros=1000, end_timestamp_micros=2000)
>>>
>>> d1 = {field.name: value for field, value in t1.ListFields()}
>>> d2 = {field.name: value for field, value in t2.ListFields()}
>>> d3 = {field.name: value for field, value in t3.ListFields()}
>>> d4 = {field.name: value for field, value in t4.ListFields()}
>>>
>>> d1
{}
>>> d2
{'start_timestamp_micros': 100}
>>> d3
{'end_timestamp_micros': 200}
>>> d4
{'end_timestamp_micros': 2000, 'start_timestamp_micros': 1000}
```

(Probably not better.)